### PR TITLE
connectors-ci: better modified connectors detection logic

### LINKED ISSUE #28419: Java connectors are being tested when running unrelated connector
## Overview 
This is not fully confirmed but there has been a report of
1. When running test in ci on `source-pokeapi` that unrelated java connectors were already run
2. When running test locally for a python connector, you will see a list of java connectors. they are not run but the list is concerning.

### LINKED ISSUE #28856: connectors-ci: ExecuteTimeoutError: Request timed out. Try setting a higher value in 'execute_timeout' config for this `dagger.Connec...
Sentry Issue: [CONNECTORS-CI-S](https://airbytehq.sentry.io/issues/4345311552/?referrer=github_integration)

```
CancelledError: 
  File "httpcore/_backends/anyio.py", line 34, in read
    return await self._stream.receive(max_bytes=max_bytes)
  File "anyio/_backends/_asyncio.py", line 1203, in receive
    await self._protocol.read_event.wait()
  File "asyncio/locks.py", line 214, in wait
    await fut

TimeoutError: 
  File "httpcore/_exceptions.py", line 10, in map_exceptions
    yield
  File "httpcore/_backends/anyio.py", line 32, in read
    with anyio.fail_after(timeout):
  File "anyio/_core/_tasks.py", line 119, in __exit__
    raise TimeoutError

ReadTimeout: 
(7 additional frame(s) were not displayed)
...
  File "httpcore/_async/http11.py", line 164, in _receive_response_headers
    event = await self._receive_event(timeout=timeout)
  File "httpcore/_async/http11.py", line 200, in _receive_event
    data = await self._network_stream.read(
  File "httpcore/_backends/anyio.py", line 31, in read
    with map_exceptions(exc_map):
  File "contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "httpcore/_exceptions.py", line 14, in map_exceptions
    raise to_exc(exc) from exc

ReadTimeout: 
(10 additional frame(s) were not displayed)
...
  File "httpx/_client.py", line 1682, in _send_handling_redirects
    response = await self._send_single_request(request)
  File "httpx/_client.py", line 1719, in _send_single_request
    response = await transport.handle_async_request(request)
  File "httpx/_transports/default.py", line 352, in handle_async_request
    with map_httpcore_exceptions():
  File "contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "httpx/_transports/default.py", line 77, in map_httpcore_exceptions
    raise mapped_exc(message) from exc

ExecuteTimeoutError: Request timed out. Try setting a higher value in 'execute_timeout' config for this `dagger.Connection()`.
(13 additional frame(s) were not displayed)
...
  File "pipelines/utils.py", line 478, in invoke
    pipeline_success = super().invoke(ctx)
  File "pipelines/commands/groups/connectors.py", line 352, in publish
    publish_connector_contexts = anyio.run(
  File "pipelines/pipelines/connectors.py", line 90, in run_connectors_pipelines
    async with anyio.create_task_group() as tg_main:
  File "<@beartype(dagger.api.gen.Container.sync) at 0x7fa32795e170>", line 11, in sync

Dagger Command publish failed
```

### PR DESCRIPTION
**Pull request by alafanechere**
## What
Closes https://github.com/airbytehq/airbyte/issues/28856
Closes #28419

Bug: If a merged PR changes one metadata.yaml but the PR potentially touches multiple connector due to dependency update or other connector change without version bump the `publish` workflow will publish all the connectors, leading to long, unexpected publish workflow.

Solution: Only publish connectors for which a metadata.yaml file was updated.

## How
When `--modified` is used in the publish command, we only keep connectors that have a modified metadata.yaml file.
This required changing the logic how connectors are selected by command (`test` / `publish` / `build`).

All the connectors command share the same way of selecting connectors by name, language or release stage via the `connectors` command group.
How connectors get added to the `selected_connectors_and_file` list is a per command logic (`test` / `publish` / `build`)

## Recommended reading order
https://github.com/airbytehq/airbyte/blob/0238aca3bfb006db73e58eaa1932914914388a8e/airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py#L308
https://github.com/airbytehq/airbyte/blob/aabbf8c0a34b47d13859d996662f921ca38cead6/airbyte-ci/connectors/pipelines/tests/test_commands/test_groups/test_connectors.py#L37

### CODE REVIEW

File: airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py [hunk 1]
```diff
@@ -394,23 +390,18 @@ def list(
...
-        # We only want to format the connector that with modified files on the current branch.
-        connectors_and_files_to_format = [
-            (connector, modified_files) for connector, modified_files in ctx.obj["selected_connectors_and_files"].items() if modified_files
-        ]
-    else:
-        # We explicitly want to format specific connectors
-        connectors_and_files_to_format = [
-            (connector, modified_files) for connector, modified_files in ctx.obj["selected_connectors_and_files"].items()
-        ]
+def format_code(ctx: click.Context) -> bool:
```
Comment (alafanechere) [inline airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py:395]: small renaming, `format` is a reserved function name in python.

File: airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py [hunk 2]
```diff
@@ -87,45 +87,26 @@ def connectors(
...
-    if release_stages:
-        selected_connectors_and_files = {
-            connector: selected_connectors_and_files[connector]
-            for connector in selected_connectors_and_files
-            if connector.release_stage in release_stages
-        }
-
-    ctx.obj["selected_connectors_and_files"] = selected_connectors_and_files
-    ctx.obj["selected_connectors_names"] = [c.technical_name for c in selected_connectors_and_files.keys()]
+    ctx.obj["all_connectors"] = get_all_released_connectors()
```
Comment (pedroslopez) [inline airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py:96]: nit: the `get_all_released_connectors()` name is a bit confusing. Reading the name and seeing `released` I would expect this to only have things that are published, but I _think_ it returns all connectors.

What does `released` refer to here?

File: airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py [hunk 3]
```diff
@@ -142,7 +123,13 @@ def test(
         main_logger.info("Skipping connectors tests for draft pull request.")
         sys.exit(0)
 
-    main_logger.info(f"Will run the test pipeline for the following connectors: {', '.join(ctx.obj['selected_connectors_names'])}")
+    if ctx.obj["select_modified_connectors"]:
+        ctx.obj["selected_connectors_and_files"] = {
+            **ctx.obj["selected_connectors_and_files"],
+            **get_modified_connectors_and_files(ctx.obj["modified_files"]),
+        }
```
Comment (pedroslopez) [inline airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py:130]: Can you help me understand what's going on here with the `**`? 

I think this means "merge `ctx.obj["selected_connectors_and_files"]` with `get_modified_connectors_and_files(ctx.obj["modified_files"])`, but shouldn't `select_modified_connectors` only have the result of `get_modified_connectors_and_files`? 

It seems like this is adding what you've selected with `--name` or other args with the modified connectors. But I would expect the `--name`, `--language`, etc to _filter_ the modified connectors. At least that was the way it was working before and is how I stopped java connectors from publishing here https://github.com/airbytehq/airbyte/pull/28344

**Review by pedroslopez — CHANGES_REQUESTED** [review]
Agree with the general approach of only considering metadata changes when publishing, however, I think there's a general change in behavior that's going on here.

Previously, the `--modified`, `--name`, `--language`, etc flags were working as an `and`. So if you have `--modified` and `--language`, it would only affect modified connectors whose language is as specified. Now, it's working as an `or`, so the same combination of args would affect modified connectors, as well as any connector with the specified language.

IMO I think the `and` is more useful and is how we implemented the stopgap for java connector publishing in https://github.com/airbytehq/airbyte/pull/28344. Please let me know if there's something I'm missing!

File: airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py [hunk 4]
```diff
@@ -142,7 +123,13 @@ def test(
...
+    main_logger.info(ctx.obj)
```
[the remark this answers is not in the document]
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py:131]: Thanks for spotting this. A debugging leftover...

File: airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py (hunk 3 above)
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py:130]: Yep `**` expands a dictionnary, in this context it indeeds merge two dictonaries into one. 
And yes you're right it slightly changes the logic. I'll try to revert to the original.
The expected logic is:
* Combination of same options like `--name`, `--language` etc. act as `OR`
* Combination of different options are `AND` `--language=python --language=java --release-stage=generally_available` should only test Python and java GA connectors.

I'll make sure to add a test for this.

File: airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py:159]: Think we still need to remove this guy! 😅

**Review by bnchrch — APPROVED** [review]
LGTM!

File: airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py (hunk 3 above)
Comment (pedroslopez) [reply airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py:158]: @alafanechere I think we're still `and`-ing modified and the other filters. Why does --modified not work like the others? 

For example `--modified --language=java` would select only the modified java connectors. Maybe I'm not reading this right...

Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py]: @pedroslopez You're right again 😮‍💨 I originally thought it was OK to `or` the modified connectors but for the publish use case is was cool to just bypass specific connector language.
I changed the code structure again to centralize the logic even more and added thorough unit tests.

**Comment by octavia-squidington-iii** [comment]
##  source-openweather test report (commit [`0de7a484fd`](https://github.com/airbytehq/airbyte/pull/28855/commits/0de7a484fd2d63e7032fd20dabbb60e4a540ff86)) - ❌

⏲️  Total pipeline duration: 19.37s 

| Step   | Result   |
|--------|----------|

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci-internal/connectors/test/pull_request/augustin_connectors-ci_better-modified-connectors-detection/1690975314/0de7a484fd2d63e7032fd20dabbb60e4a540ff86/source-openweather/0.1.6/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-openweather test
```

**Comment by octavia-squidington-iii** [comment]
##  source-openweather test report (commit [`1da9809dc9`](https://github.com/airbytehq/airbyte/pull/28855/commits/1da9809dc9777bcbfc2186129b400a65ce119358)) - ❌

⏲️  Total pipeline duration: 16.79s 

| Step   | Result   |
|--------|----------|

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci-internal/connectors/test/pull_request/augustin_connectors-ci_better-modified-connectors-detection/1690975473/1da9809dc9777bcbfc2186129b400a65ce119358/source-openweather/0.1.6/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-openweather test
```

**Comment by octavia-squidington-iii** [comment]
##  source-openweather test report (commit [`c181cf6cad`](https://github.com/airbytehq/airbyte/pull/28855/commits/c181cf6cadf19b1ad2f0d5d7e0d9914f8f68d889)) - ❌

⏲️  Total pipeline duration: 02mn24s 

| Step                                                                      | Result   |
|:--------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-openweather/metadata.yaml | ✅        |
| Connector version semver check                                            | ✅        |
| QA checks                                                                 | ✅        |
| Code format checks                                                        | ✅        |
| Connector package install                                                 | ✅        |
| Build source-openweather docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                | ✅        |
| Acceptance tests                                                          | ❌        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci-internal/connectors/test/pull_request/augustin_connectors-ci_better-modified-connectors-detection/1690976510/c181cf6cadf19b1ad2f0d5d7e0d9914f8f68d889/source-openweather/0.1.6/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-openweather test
```

**Comment by octavia-squidington-iii** [comment]
##  source-openweather test report (commit [`351725fd0c`](https://github.com/airbytehq/airbyte/pull/28855/commits/351725fd0c7c61e01fdbff01024434d5599b079a)) - ❌

⏲️  Total pipeline duration: 02mn42s 

| Step                                                                      | Result   |
|:--------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-openweather/metadata.yaml | ✅        |
| Connector version semver check                                            | ✅        |
| QA checks                                                                 | ✅        |
| Code format checks                                                        | ✅        |
| Connector package install                                                 | ✅        |
| Build source-openweather docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                | ✅        |
| Acceptance tests                                                          | ❌        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci-internal/connectors/test/pull_request/augustin_connectors-ci_better-modified-connectors-detection/1690976885/351725fd0c7c61e01fdbff01024434d5599b079a/source-openweather/0.1.6/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-openweather test
```

**Comment by octavia-squidington-iii** [comment]
##  source-faker test report (commit [`351725fd0c`](https://github.com/airbytehq/airbyte/pull/28855/commits/351725fd0c7c61e01fdbff01024434d5599b079a)) - ❌

⏲️  Total pipeline duration: 02mn49s 

| Step                                                                | Result   |
|:--------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-faker/metadata.yaml | ✅        |
| Connector version semver check                                      | ✅        |
| QA checks                                                           | ✅        |
| Code format checks                                                  | ✅        |
| Connector package install                                           | ✅        |
| Build source-faker docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                          | ✅        |
| Acceptance tests                                                    | ❌        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci-internal/connectors/test/pull_request/augustin_connectors-ci_better-modified-connectors-detection/1690976885/351725fd0c7c61e01fdbff01024434d5599b079a/source-faker/4.0.0/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-faker test
```

**Comment by octavia-squidington-iii** [comment]
##  source-openweather test report (commit [`a2fe34bbdc`](https://github.com/airbytehq/airbyte/pull/28855/commits/a2fe34bbdce181b031d96f0aafb2b560c9d56173)) - ✅

⏲️  Total pipeline duration: 02mn23s 

| Step                                                                      | Result   |
|:--------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-openweather/metadata.yaml | ✅        |
| Connector version semver check                                            | ✅        |
| QA checks                                                                 | ✅        |
| Code format checks                                                        | ✅        |
| Connector package install                                                 | ✅        |
| Build source-openweather docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                | ✅        |
| Acceptance tests                                                          | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci-internal/connectors/test/pull_request/augustin_connectors-ci_better-modified-connectors-detection/1690986023/a2fe34bbdce181b031d96f0aafb2b560c9d56173/source-openweather/0.1.6/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-openweather test
```

**Comment by octavia-squidington-iii** [comment]
##  source-faker test report (commit [`a2fe34bbdc`](https://github.com/airbytehq/airbyte/pull/28855/commits/a2fe34bbdce181b031d96f0aafb2b560c9d56173)) - ✅

⏲️  Total pipeline duration: 02mn28s 

| Step                                                                | Result   |
|:--------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-faker/metadata.yaml | ✅        |
| Connector version semver check                                      | ✅        |
| QA checks                                                           | ✅        |
| Code format checks                                                  | ✅        |
| Connector package install                                           | ✅        |
| Build source-faker docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                          | ✅        |
| Acceptance tests                                                    | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci-internal/connectors/test/pull_request/augustin_connectors-ci_better-modified-connectors-detection/1690986023/a2fe34bbdce181b031d96f0aafb2b560c9d56173/source-faker/4.0.0/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-faker test
```

**Review by pedroslopez — APPROVED** [review]
giphy

Some comments, but looks good! Thank you 😄

File: airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py [hunk 5]
```diff
@@ -47,6 +48,72 @@ def validate_environment(is_local: bool, use_remote_secrets: bool):
...
+    all_connectors = get_all_connectors_in_repo()
+
+    if metadata_changes_only and not modified:
+        main_logger.info("--metadata-changes-only overrides --modified")
+        modified = True
+
+    selected_modified_connectors = get_modified_connectors(modified_files) if modified else set()
+    selected_connectors_by_name = validate_selected_connectors_exist(
+        {Connector(technical_name=name) for name in selected_names}, all_connectors
+    )
```
[the remark this answers is not in the document]
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py]: I changed it for consistency. Connector name validation is now happening at click option parsing [26a68e5](https://github.com/airbytehq/airbyte/pull/28855/commits/26a68e56656a915a97ff5c792ffdf3395434d73d)

File: airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py [hunk 6]
```diff
@@ -87,45 +161,12 @@ def connectors(
 
     ctx.ensure_object(dict)
     ctx.obj["use_remote_secrets"] = use_remote_secrets
-    ctx.obj["connector_names"] = names
-    ctx.obj["connector_languages"] = languages
-    ctx.obj["release_states"] = release_stages
-    ctx.obj["modified"] = modified
```
[the remark this answers is not in the document]
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/commands/groups/connectors.py:93]: I don't think they're used elsewhere. Will confirm by running a pre-release. I'm not against re-adding these if needed but ATM I believe keeping `ctx.obj` with used fields is better for readability / debugging.

**Pull request merged**